src/ucp/rma/amo_basic.c

```diff
@@ -0,0 +1,351 @@
+    req->send.amo.remote_addr = remote_addr;
+    req->send.amo.rkey        = rkey;
+    req->send.amo.value       = value;
+#if ENABLE_ASSERT
+    req->send.lane            = UCP_NULL_LANE;
+#endif
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_amo_init_fetch(ucp_request_t *req, ucp_ep_h ep, void *buffer,
+                   uct_atomic_op_t op, size_t op_size, uint64_t remote_addr,
+                   ucp_rkey_h rkey, uint64_t value)
+{
+    ucp_amo_init_common(req, ep, op, remote_addr, rkey, value, op_size);
+    req->send.state.uct_comp.count  = 1;
+    req->send.state.uct_comp.func   = ucp_amo_completed_single;
+    req->send.uct.func              = ucp_amo_basic_progress_fetch;
+    req->send.buffer                = buffer;
+}
+
+static UCS_F_ALWAYS_INLINE
+void ucp_amo_init_post(ucp_request_t *req, ucp_ep_h ep, uct_atomic_op_t op,
+                       size_t op_size, uint64_t remote_addr, ucp_rkey_h rkey,
+                       uint64_t value)
+{
+    ucp_amo_init_common(req, ep, op, remote_addr, rkey, value, op_size);
+    req->send.uct.func              = ucp_amo_basic_progress_post;
+}
+
+ucs_status_ptr_t ucp_atomic_fetch_nb(ucp_ep_h ep, ucp_atomic_fetch_op_t opcode,
+                                     uint64_t value, void *result, size_t op_size,
+                                     uint64_t remote_addr, ucp_rkey_h rkey,
+                                     ucp_send_callback_t cb)
+{
+    ucs_status_ptr_t status_p;
+    ucs_status_t status;
+    ucp_request_t *req;
+
+    UCP_AMO_CHECK_PARAM(remote_addr, op_size, opcode, UCP_ATOMIC_FETCH_OP_LAST,
+                        UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM));
+
+    UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(ep->worker);
+
+    status = UCP_RKEY_RESOLVE(rkey, ep, amo);
+    if (status != UCS_OK) {
+        status_p = UCS_STATUS_PTR(UCS_ERR_UNREACHABLE);
+        goto out;
+    }
+
+    req = ucp_request_get(ep->worker);
+    if (ucs_unlikely(NULL == req)) {
+        status_p = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
+        goto out;
+    }
+
+    ucp_amo_init_fetch(req, ep, result, ucp_uct_fop_table[opcode], op_size,
+                       remote_addr, rkey, value);
+
+    status_p = ucp_rma_send_request_cb(req, cb);
+
+out:
+    UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(ep->worker);
+    return status_p;
+}
+
+ucs_status_t ucp_atomic_post(ucp_ep_h ep, ucp_atomic_post_op_t opcode, uint64_t value,
+                             size_t op_size, uint64_t remote_addr, ucp_rkey_h rkey)
+{
+    ucs_status_ptr_t status_p;
+    ucs_status_t status;
+    ucp_request_t *req;
+    uct_atomic_op_t op;
+
+    UCP_AMO_CHECK_PARAM(remote_addr, op_size, opcode, UCP_ATOMIC_POST_OP_LAST,
+                        UCS_ERR_INVALID_PARAM);
+
+    UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(ep->worker);
+
+    op     = ucp_uct_op_table[opcode];
+    status = UCP_RKEY_RESOLVE(rkey, ep, amo);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    req = ucp_request_get(ep->worker);
```

Comment:
why do we need a request if it is not exposed in API?

Reply:
we need something to progress (same as with ucp_put_nbi, for example)